### PR TITLE
JIRA-OSDOCS3037: Developed content for persistent storage using AWS EBS

### PR DESCRIPTION
**Pull request by sayjadha**
This PR is to address JIRA issue: https://issues.redhat.com/browse/OSDOCS-3037

Topic updated: https://deploy-preview-42174--osdocs.netlify.app/openshift-rosa/latest/storage/persistent_storage/rosa-persistent-storage-aws-ebs.html

The exact changes are as follows:

- Developed a new topic 'Persistent storage using AWS Elastic Block Store (EBS)'
- Updated the topic map topic_map_rosa.yml to include the newly-created topic. 

Repo: Request cherrypick to enterprise-4.9 and enterprise-4.10

### CODE REVIEW

**Comment by sayjadha** [comment]
To confirm with QE: Does a ROSA cluster support 39 or 25 EBS volumes attached to one node? It says 25 here: [https://kubernetes.io/docs/concepts/storage/storage-limits/#dynamic-volume-limits](https://www.google.com/url?q=https://kubernetes.io/docs/concepts/storage/storage-limits/%23dynamic-volume-limits&sa=D&source=docs&ust=1645409413945668&usg=AOvVaw0ythlc15gOmvb3zX-1XqlG)

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+You must use either in-tree or CSI volumes and their respective storage classes, but never both volume types at the same time. The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.
+====
+
+== Creating a persistent volume claim
+
+.Prerequisites
+Storage must exist in the underlying infrastructure before it can be mounted as a volume in the ROSA cluster.
+
+.Procedure
+. In the OpenShift Cluster Manager (OCM) console, click *Storage → Persistent Volume Claims*.
```
Comment (yuwang-RH) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:60]: It should be OpenShift cluster console rather than 'OpenShift Cluster Manager (OCM) console'.

**Comment by yuwang-RH** [comment]
@sayjadha Leave one comment, other LGTM.

File: _topic_maps/_topic_map_rosa.yml [hunk 2]
```diff
@@ -205,6 +205,8 @@ Topics:
   Topics:
     - Name: Persistent storage using AWS EFS
       File: osd-persistent-storage-aws
+    - Name: Persistent storage using AWS Elastic Book Store (EBS)
```
Comment (jeana-redhat) [inline _topic_maps/_topic_map_rosa.yml:208]:
```suggestion
    - Name: Persistent storage using AWS Elastic Block Store (EBS)
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+== Capacity of EBS volumes on a node
+By default, a ROSA cluster supports a maximum of 39 EBS volumes attached to one node. This limit is consistent with the link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits[AWS volume limits]. The volume limit depends on the instance type.
+
+[IMPORTANT]
+====
+You must use either in-tree or CSI volumes and their respective storage classes, but never both volume types at the same time. The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.
+====
+
+== Creating a persistent volume claim
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:54]: This should be its own procedure module.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+====
+ROSA defaults to using an in-tree or non-Container Storage Interface (CSI) plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.
+====
+
+[IMPORTANT]
+====
+High-availability of storage in the infrastructure is left to the underlying storage provider.
+====
+
+== Format of persistent volumes
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:41]: This may be different for ROSA docs, but OCP core would require an [id] for each section. Same applies for `== Capacity of EBS volumes on a node`

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+[IMPORTANT]
+====
+ROSA defaults to using an in-tree or non-Container Storage Interface (CSI) plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.
+====
+
+[IMPORTANT]
+====
+High-availability of storage in the infrastructure is left to the underlying storage provider.
+====
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:39]: Is it possible to restructure to avoid stacked admonitions? If not, consider including them both in one box as a bulleted list.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+. In the OpenShift Cluster Manager (OCM) console, click *Storage → Persistent Volume Claims*.
+. In the persistent volume claims overview, click *Create Persistent Volume Claim*.
+. Define the desired options on the page that appears.
+  .. Select the storage class created previously from the drop-down menu.
+  .. Enter a unique name for the storage claim.
+  .. Select the access mode. This selection determines the read and write access for the newly created storage claim.
+  .. Define the size of the storage claim.
+. Click *Create* to create the persistent volume claim and generate a persistent volume.
+
+== Additional resources
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:69]: Check this against https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#additional-resources-sections
Needs some tagging, should also be a list not a paragraph.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+== Creating a persistent volume claim
+
+.Prerequisites
+Storage must exist in the underlying infrastructure before it can be mounted as a volume in the ROSA cluster.
+
+.Procedure
+. In the OpenShift Cluster Manager (OCM) console, click *Storage → Persistent Volume Claims*.
+. In the persistent volume claims overview, click *Create Persistent Volume Claim*.
+. Define the desired options on the page that appears.
+  .. Select the storage class created previously from the drop-down menu.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:63]:
```suggestion
  .. Select the previously created storage class from the drop-down menu.
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  .. Enter a unique name for the storage claim.
+  .. Select the access mode. This selection determines the read and write access for the newly created storage claim.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:65]:
```suggestion
  .. Select the access mode. This selection determines the read and write access for the storage claim.
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 9]
```diff
@@ -0,0 +1,70 @@
+:_content-type: ASSEMBLY
+[id="rosa-persistent-storage-aws-ebs"]
+= Persistent storage using AWS Elastic Block Store (EBS)
+include::modules/attributes-openshift-dedicated.adoc[]
+:context: rosa-persistent-storage-aws-ebs
+
+toc::[]
+
+{product-title} (ROSA) clusters are pre-built with two storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:9]:
```suggestion
{product-title} (ROSA) clusters are prebuilt with two storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+Following are the two pre-built storage classes:
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:11]:
```suggestion
Following are the two prebuilt storage classes:
```
This is a sentence fragment, but you could consider just making it a table title (using a `.Heading`) to avoid having to say too much :)

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+| gp2 | kubernetes.io/aws-ebs
+
+| gp2-csi | ebs.csi.aws.com
+
+|===
+
+[NOTE]
+====
+The gp2 storage class is set as default; however, you can select either one as the default storage class.
+====
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:27]: Consider just adding `(default)` next to the `gp2` line in the table, or possibly just making this a regular sentence in one of the paragraphs.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+The Kubernetes persistent volume framework enables the administrators to provision a cluster with persistent storage and gives users a way to request those resources without having any knowledge of the underlying infrastructure. You can dynamically provision the AWS EBS volumes. Persistent volumes are not bound to a single project or namespace; therefore, the volumes can be shared across ROSA clusters. Persistent volume claims are specific to a project or namespace and can be requested by users.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:29]:
```suggestion
The Kubernetes persistent volume framework enables administrators to provision a cluster with persistent storage and gives users a way to request those resources without having any knowledge of the underlying infrastructure. You can dynamically provision AWS EBS volumes. Persistent volumes are not bound to a single project or namespace; therefore, the volumes can be shared across ROSA clusters. Persistent volume claims are specific to a project or namespace and can be requested by users.
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+[IMPORTANT]
+====
+ROSA defaults to using an in-tree or non-Container Storage Interface (CSI) plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:33]: This suggestion applies if "in-tree" = non-CSI (I think it does)
```suggestion
ROSA defaults to using an in-tree, or non-Container Storage Interface (CSI), plug-in to provision AWS EBS storage. In future ROSA versions, volumes provisioned using existing in-tree plug-ins are planned for migration to their equivalent CSI driver. After full migration, the in-tree plug-ins will eventually be removed from the future versions of ROSA.
```
Can you rephrase that final sentence to avoid "will" (definite) and speak more about "plans"?

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+====
+
+[IMPORTANT]
+====
+High-availability of storage in the infrastructure is left to the underlying storage provider.
+====
+
+== Format of persistent volumes
+Before a ROSA cluster mounts the volume and passes it to a container, the cluster checks that the volume contains a file system as specified by the *fsType* parameter in the persistent volume definition. If the device is not formatted with the file system, all data from the device is erased and the device is automatically formatted with the given file system.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:42]:
```suggestion
Before a ROSA cluster mounts the volume and passes it to a container, the cluster checks that the volume contains a file system as specified by the *fsType* parameter in the persistent volume definition. If the device is not formatted with a file system, all data from the device is erased and the device is automatically formatted with the given file system.
```

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+This verification enables you to use unformatted AWS volumes as persistent volumes, as the ROSA cluster formats the AWS volumes before the first use.
+
+
+== Capacity of EBS volumes on a node
+By default, a ROSA cluster supports a maximum of 39 EBS volumes attached to one node. This limit is consistent with the link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html#linux-specific-volume-limits[AWS volume limits]. The volume limit depends on the instance type.
+
+[IMPORTANT]
+====
+You must use either in-tree or CSI volumes and their respective storage classes, but never both volume types at the same time. The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.
```
Comment (jeana-redhat) [inline storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:51]: > The maximum attached EBS volume number is counted separately for in-tree and CSI volumes.

What does this mean? you could have up to 39 of each type?

**Review by jeana-redhat — COMMENTED** [review]
Just a few very minor style suggestions, and couple formatting issues (needs a module for the procedure, and there are some late-breaking changes to how we do "Additional resources" - see link in comment).

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 1 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:60]: Made the change, thanks.

File: _topic_maps/_topic_map_rosa.yml (hunk 2 above)
Comment (sayjadha) [reply _topic_maps/_topic_map_rosa.yml:208]: Made the change, thanks.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 6 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:69]: Thanks, made changes as necessary.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 7 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:63]: Made the change.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 8 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:65]: Made the change.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 9 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:9]: Made the change.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 10 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:11]: Made the change.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 11 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:27]: Moved the text into a regular sentence in a paragraph below the table.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 12 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:29]: Made the change.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 13 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:33]: Made the edit.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 14 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:42]: Made the edit.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 15 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:51]: That's right. Made the change to clarify this.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 5 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:39]: Thanks! Added both notes in one box as a bulleted list.

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 4 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:41]: Added the [id] for each section, and I'll try using this as basic repo hygiene going forward. Thank you!

File: storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc (hunk 3 above)
Comment (sayjadha) [reply storage/persistent_storage/rosa-persistent-storage-aws-ebs.adoc:54]: Done

**Comment by sayjadha** [comment]
Thank you for the review, @jeana-redhat. Your feedback is incorporated. Awaiting PM review before we can merge this PR. Will keep you posted, thanks!

**Comment by 0kashi** [comment]
lgtm - can we change the order in the left menu though to have EBS appear before EFS?

**Comment by sayjadha** [comment]
@0kashi: Your feedback is addressed. Thank you.
@jeana-redhat: I'll check with Kathryn about the text beside the URL in 'Additional Resources' section. Thank you for your help!

**Comment by sayjadha** [comment]
@jeana-redhat, I spoke with Kathryn and they said the text beside the URL should be removed. They suggested that I can turn that additional resources section into a note or body text. I've moved the URL 'Elastic Block Store CSI Driver Operator' moved to an appropriate place in the topic.
The PM review is completed too. Could you please help merge the issue? Thank you.
